File: backend/common/money.py
"""
Money math shared by quotes and invoices. Everything is Decimal, never
float, and always quantized to cents. VAT is the flat South African rate
(currently 15%) applied to (subtotal - discount) only when the document is
marked VAT-applicable.
"""

from decimal import ROUND_HALF_UP, Decimal
from typing import Iterable

CENTS = Decimal("0.01")
VAT_RATE = Decimal("0.15")


def quantize(value) -> Decimal:
    return Decimal(value).quantize(CENTS, rounding=ROUND_HALF_UP)


def compute_line_total(quantity, unit_price) -> Decimal:
    return quantize(Decimal(quantity) * Decimal(unit_price))


def compute_document_totals(
    line_totals: Iterable[Decimal], discount_amount, is_vat_applicable: bool
) -> tuple[Decimal, Decimal, Decimal]:
    """Returns (subtotal, vat_amount, total)."""
    subtotal = quantize(sum((Decimal(t) for t in line_totals), Decimal("0")))
    discount = quantize(Decimal(discount_amount or 0))
    taxable = subtotal - discount
    if taxable < 0:
        taxable = Decimal("0.00")
    vat_amount = quantize(taxable * VAT_RATE) if is_vat_applicable else Decimal("0.00")
    total = quantize(taxable + vat_amount)
    return subtotal, vat_amount, total

File: android/core-domain/src/main/kotlin/com/ops/coredomain/Money.kt
package com.ops.coredomain

import java.math.BigDecimal
import java.math.RoundingMode

/**
 * Money math shared by quotes and invoices. Mirrors backend/common/money.py
 * EXACTLY — same rounding mode, same order of operations — because the
 * Android app computes these totals locally for instant offline UI, and the
 * numbers must match what the server recomputes as canonical on sync. See
 * docs/API_CONTRACT.md ("subtotal/vat_amount/total ... are always recomputed
 * server-side ... a client may compute them locally for instant offline UI").
 *
 * Everything is [BigDecimal], never Double/Float — money is never a binary
 * float in this codebase, on either side of the wire.
 */
object Money {

    /** Current flat South African VAT rate. Server-configurable per business
     * only via the is_vat_applicable flag on a document, not a variable rate. */
    val VAT_RATE: BigDecimal = BigDecimal("0.15")

    private val CENTS = BigDecimal("0.01")
    private val ZERO = BigDecimal("0.00")

    /** Quantize to 2 decimal places, half-up — same as Python's
     * `Decimal(value).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)`. */
    fun quantize(value: BigDecimal): BigDecimal = value.setScale(2, RoundingMode.HALF_UP)

    /** A single quote/invoice line item's total: quantity * unit price, quantized to cents. */
    fun computeLineTotal(quantity: BigDecimal, unitPrice: BigDecimal): BigDecimal =
        quantize(quantity.multiply(unitPrice))

    /**
     * Whole-document totals from its line items. Discount is applied to the
     * subtotal BEFORE VAT; the taxable amount is never allowed to go
     * negative (a discount larger than the subtotal simply zeroes out the
     * taxable amount, it does not produce a negative VAT/total). VAT is a
     * flat 0.00 when the document is not VAT-applicable at all.
     */
    fun computeDocumentTotals(
        lineTotals: List<BigDecimal>,
        discountAmount: BigDecimal,
        isVatApplicable: Boolean,
    ): DocumentTotals {
        val subtotal = quantize(lineTotals.fold(BigDecimal.ZERO) { acc, t -> acc.add(t) })
        val discount = quantize(discountAmount)
        var taxable = subtotal.subtract(discount)
        if (taxable.signum() < 0) {
            taxable = ZERO
        }
        val vatAmount = if (isVatApplicable) quantize(taxable.multiply(VAT_RATE)) else ZERO
        val total = quantize(taxable.add(vatAmount))
        return DocumentTotals(subtotal = subtotal, vatAmount = vatAmount, total = total)
    }
}

/** subtotal / vatAmount / total for a quote or invoice, all quantized to 2dp. */
data class DocumentTotals(
    val subtotal: BigDecimal,
    val vatAmount: BigDecimal,
    val total: BigDecimal,
)
